### PR TITLE
Implement `list_sizes()`

### PR DESCRIPTION
**Pull request by romainfrancois**
`vec_sizes()` is to `vec_size()` what `lengths()` is to `length()`: 

``` r
library(vctrs)

data <- list(iris, mtcars, 1:10)
vec_sizes(data)
#> [1] 150  32  10
```

The 😈 plan is to use this in a vctrs based implementation of things like `dplyr::summarise()` and `dplyr::mutate()` e.g. evaluate the expression in each group and then assert that the sizes are good (all equal to 1 if summaries, equal to the group sizes on a mutate). 

The PR is wip, if we want that I'll add unit tests, etc ...

### CODE REVIEW

**Comment by romainfrancois** [comment]
Alternatively or additionally, it would be nice to have a C api for `vec_size()` that we can use from internals of e.g. `dplyr`.

**Comment by DavisVaughan** [comment]
Not sure if it matters, but it looks like this would also work on a data frame and return the size of each column (all the same size)?

**Comment by lionel-** [comment]
I think it'd be better to export `vec_size()` as a C callable, if needed.

> Not sure if it matters, but it looks like this would also work on a data frame and return the size of each column (all the same size)?

That doesn't seem appropriate. In general this function seems ill-suited for vctrs at the moment because it's about `[[`.

**Comment by romainfrancois** [comment]
Sure. I can live with `map_int(vec_size)` for now, having `vec_size()` C callable means I can implement something that tests is all the sizes are 1 instead of getting all the sizes and then do a `any(sizes != 1)`

**Comment by lionel-** [comment]
Let's revisit once we have figured out `vec_slice2()` and the `tbl_` API.

**Comment by lionel-** [comment]
From a discussion with @DavisVaughan, we could have `list_sizes()` that fails with atomic vectors (including dfs).

**Comment by DavisVaughan** [comment]
@romainfrancois I'll take over this PR if that's alright with you

**Comment by romainfrancois** [comment]
@DavisVaughan sure. `dplyr` still has a `dplyr_vec_sizes` but is unused at this point.

**Comment by DavisVaughan** [comment]
Pretty nice improvement

``` r
library(purrr)
library(vctrs)

x <- rep_len(list(1:5 + 0L), 100)

bench::mark(
  map_int(x, vec_size),
  list_sizes(x)
)
#> # A tibble: 2 x 6
#>   expression                min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr>           <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 map_int(x, vec_size)  63.34µs  77.62µs    12802.    68.6KB     91.8
#> 2 list_sizes(x)          2.81µs   3.08µs   287798.     4.3KB      0

list_sizes(mtcars)
#> Error: `x` must be a list.
```

<sup>Created on 2020-03-18 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

File: NEWS.md [hunk 1]
```diff
@@ -1,6 +1,9 @@
 
 # vctrs (development version)
 
+* New `list_sizes()` for computing the size of every element in a list.
+  `list_sizes()` is to `vec_size()` as `lengths()` is to `length()`.
```
Comment (lionel-) [inline NEWS.md]: except that it only supports lists. Atomic vectors and data frames cause an error.

File: src/size.c [hunk 2]
```diff
@@ -98,6 +98,30 @@ SEXP vctrs_size(SEXP x) {
...
+    SEXP elt = VECTOR_ELT(x, i);
+    p_out[i] = vec_size(elt);
+  }
+
+  UNPROTECT(1);
+  return out;
+}
+
+// [[ register() ]]
+SEXP vctrs_list_sizes(SEXP x) {
```
Comment (lionel-) [inline src/size.c]: Minor: Do we need a registered alias since they have the same signature?

Comment (DavisVaughan) [reply src/size.c]: I tend to try and always keep a separate function prefixed with `vctrs_` that gets exported, whether it is strictly required or not

**Pull request merged**